### PR TITLE
Implement comment history pagination and navigation, and prevention of cross site scripting.

### PR DESCRIPTION
**Pull request by KathyFeiyang**
Implement comment history pagination and navigation, and prevention of cross site scripting:

**Comment history pagination & navigation**
- Fetch comments based on the user specified number of comments to display per page and page ID.
- Support first, last, previous and next navigation, with error handling and revised interface.
- Support "Go to page ?" to jump to a specified page, with error handling.

**Cross site scripting prevention**
- Check user comments using a Java encoder.
- Reject potentially malicious user comments and refrain from storing said comments in the backend database.
- Show a red warning below the comment submission box about the rejected comment, immediately after the comment was submitted; the warning will disappear if you refresh the page or submit a proper comment.

### CODE REVIEW

File: portfolio/src/main/webapp/index.html [hunk 1]
```diff
@@ -108,13 +108,23 @@ <h3 id="contact_me">Contact Me</h3>
...
+        <input type="number" id="max-comments-to-display" name="max-comments-to-display" onchange="addComments(1)"
+               min="0" max="" value="">
+        <p>Currently on page <b id="current-page-id"></b>. (The latest comments are displayed on top.)</p>
+        <button id="first" onclick="addComments(this.id)">&lt;&lt;</button>
+        <button id="prev" onclick="addComments(this.id)">&lt;</button>
+        <button id="next" onclick="addComments(this.id)">&gt;</button>
+        <button id="last" onclick="addComments(this.id)">&gt;&gt;</button>
+        <label for="go-to-page-id">Go to page </label>
+        <input type="number" id="go-to-page-id" name="go-to-page-id" onchange="addComments(this.value)"
+               min="0" max="" value="1">
```
Comment (kellyktan) [inline portfolio/src/main/webapp/index.html]: I would make the min here and for the max-comments-to-display 1 since there is no page "0" and setting max comments to 0 causes weirdness when trying to go to the last page.

File: portfolio/src/main/webapp/index.html [hunk 2]
```diff
@@ -108,13 +108,23 @@ <h3 id="contact_me">Contact Me</h3>
...
+        <label for="max-comments-to-display">Maximum number of entries to show on one page of the comment history:</label>
+        <input type="number" id="max-comments-to-display" name="max-comments-to-display" onchange="addComments(1)"
+               min="0" max="" value="">
+        <p>Currently on page <b id="current-page-id"></b>. (The latest comments are displayed on top.)</p>
+        <button id="first" onclick="addComments(this.id)">&lt;&lt;</button>
+        <button id="prev" onclick="addComments(this.id)">&lt;</button>
+        <button id="next" onclick="addComments(this.id)">&gt;</button>
+        <button id="last" onclick="addComments(this.id)">&gt;&gt;</button>
+        <label for="go-to-page-id">Go to page </label>
+        <input type="number" id="go-to-page-id" name="go-to-page-id" onchange="addComments(this.value)"
```
Comment (kellyktan) [inline portfolio/src/main/webapp/index.html]: I would also add some validation here around not allowing it to go over the "max number of pages".  For instance, if there is 1 page I can still enter "5" here and have it tell me I'm on page 5 lol

**Review by kellyktan — CHANGES_REQUESTED** [review]

File: portfolio/src/main/webapp/index.html (hunk 1 above)
Comment (KathyFeiyang) [reply portfolio/src/main/webapp/index.html]: Got it, I will fix this

File: portfolio/src/main/webapp/index.html (hunk 2 above)
Comment (KathyFeiyang) [reply portfolio/src/main/webapp/index.html]: HTML's max attribute only sets the biggest number users can get for using the up arrow on the right side of the text box. Users can still type in larger numbers. To do further validation, I would probably need JavaScript or backend Java code. I'll look into this.

Comment (KathyFeiyang) [reply portfolio/src/main/webapp/index.html]: Currently my code checks for this on the backend Java side. I'll try to debug that code and see if I can improve that behavior. Thanks for pointing this out! I think this is quite important

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 3]
```diff
@@ -23,28 +57,34 @@
...
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
     // Query comment history from Datastore.
     Query commentHistoryQuery = new Query("UserComment").addSort("timestamp", SortDirection.DESCENDING);
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     PreparedQuery commentHistory = datastore.prepare(commentHistoryQuery);
-    int totalComments = commentHistory.countEntities(FetchOptions.Builder.withDefaults());
+    List<Entity> commentHistoryList = commentHistory.asList(FetchOptions.Builder.withDefaults());
+    int totalComments = commentHistoryList.size();
+    int totalPages;
```
Comment (suparnabarua) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Can you move the totalPages declaration to line 99? In general it's best practice to define a variable right before it's used.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -55,24 +95,51 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
       maxCommentsToDisplay = DEFAULT_MAX_COMMENTS;
     }
 
+    // Obtain the page ID of comment history to display.
+    String pageIdParam = request.getParameter("pageId");
+    totalPages = (int) Math.ceil(((double) totalComments) / maxCommentsToDisplay);
+    try {
+      currentPageId = Integer.parseInt(pageIdParam);
```
Comment (suparnabarua) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: I think there should be a limit on the number of pages like maxPages and checks for that. Right now I can say take me to page 100 and it will say "Currently on page 100"

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -55,24 +95,51 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
+      // If the numeric input of page ID is invalid, set to 1.
+      currentPageId = Math.max(1, currentPageId);
+    } catch (NumberFormatException e) {
+      if (pageIdParam.equals("first")) {
```
Comment (suparnabarua) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Use a switch block for going through the cases for pageIdParam. For the "prev", you can do something like currentPageId = Math.max(currentPageId, currentPageId - 1) to make sure you aren't going to page 0

File: portfolio/src/main/webapp/index.html [hunk 6]
```diff
@@ -108,13 +108,23 @@ <h3 id="contact_me">Contact Me</h3>
...
+      <div id="is-latest-input-dangerous"></div>
       <p><b>Comment History</b></p>
-      <label for="maxCommentsToDisplay">Maximum number of entries to show in comment history:</label>
-      <input type="number" id="maxCommentsToDisplay" name="maxCommentsToDisplay"
-             onchange="addComments()" min="0" max="" value="">
-      <p>(The latest comments are displayed on top.)</p>
       <ul id="comment-container"></ul>
-      <button onclick="addComments()">Refresh</button>
+      <div id="pagination">
+        <label for="max-comments-to-display">Maximum number of entries to show on one page of the comment history:</label>
```
Comment (suparnabarua) [inline portfolio/src/main/webapp/index.html]: Can you break up the long lines here

**Review by suparnabarua — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 4 above)
Comment (KathyFeiyang) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Yes you're right. I'm implementing that and will submit a code change soon

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 5 above)
Comment (KathyFeiyang) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Got it, thanks!

File: portfolio/src/main/webapp/index.html (hunk 6 above)
Comment (KathyFeiyang) [reply portfolio/src/main/webapp/index.html]: This is fixed.

File: portfolio/src/main/webapp/index.html (hunk 1 above)
Comment (KathyFeiyang) [reply portfolio/src/main/webapp/index.html]: This is fixed.

File: portfolio/src/main/webapp/index.html (hunk 2 above)
Comment (KathyFeiyang) [reply portfolio/src/main/webapp/index.html]: The error checking works now, with a warning message displayed in the following 2 scenarios:
1. user enters a negative maximum number of comments to display;
2. user enters a page ID out of the range [1, total number of pages].

Implementation-wise, because I need to determine the valid range by first querying the backend database for the total number of comments, checking user input must involve the Java servlet. Therefore, I kept the error checking logic in the Java servlet. There were some bugs and missed edge cases with the code (such as with 0 max comments to display, or in situations when there is no comment/page in the database); they are now fixed. Next time, I'll definitely remember to test thoroughly.

The Java servlet is completely in charge of the logic to check user input's validity. It indicates this validity through 2 flags. The client-side code is in charge of checking those flags and showing warning messages if necessary.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 4 above)
Comment (KathyFeiyang) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: This issue is fixed, along with some other missed edges. Thanks for pointing this out!

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 7]
```diff
@@ -82,36 +84,65 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
-      maxCommentsToDisplay = Integer.parseInt(request.getParameter("maxCommentsToDisplay"));
+      maxCommentsToDisplay = Integer.parseInt(maxCommentsToDisplayStr);
       if (maxCommentsToDisplay < 0) {
         maxCommentsToDisplay = DEFAULT_MAX_COMMENTS;
+        invalidMaxComments = true;
       }
     } catch (NumberFormatException e) {
       maxCommentsToDisplay = DEFAULT_MAX_COMMENTS;
+      // Upon initialization, the field is empty; this special case doesn't indicate erroneous user input.
+      if (!maxCommentsToDisplayStr.equals("")) {
```
Comment (suparnabarua) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Can maxCommentsToDisplayStr be null? If not, switch to String.isEmpty(). Otherwise, keep it as it is.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 4 above)
Comment (suparnabarua) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Can you move this try/catch with the switch into a helper function that clearly describes the purpose? Also  you seem to override the processing here with the maxCommentsToDisplay == 0 check at the end. instead do this:
if (maxCommentsToDisplay == 0) {
      totalPages = 1;
      currentPageId = 1;
    } else {
  call helperFunction()
}

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 8]
```diff
@@ -187,16 +222,21 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
     private int defaultMaxComments;
     private int totalPages;
     private int currentPageId;
+    private boolean invalidMaxComments;
+    private boolean invalidPageId;
     private boolean isLatestInputDangerous;
 
     public commentDataPackage(List<UserComment> inputComments, int inputTotalComments,
```
Comment (suparnabarua) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Don't use camelcase for class name

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 9]
```diff
@@ -136,9 +167,13 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
 
     // Package a history of comment objects, the total number of comments, the default number of
     // comments, the total number of pages, the current page ID, and whether the latest user
-    // input was dangerous and rejected, to JSON format.
+    // input was dangerous and rejected, whether the user input values of number of comments to 
+    // display & page ID are invalid, to JSON format.
     commentDataPackage commentData = new commentDataPackage(comments, totalComments, DEFAULT_MAX_COMMENTS,
-                                                            totalPages, currentPageId, isLatestInputDangerous);
+                                                            totalPages, currentPageId, invalidMaxComments,
```
Comment (suparnabarua) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: There's too many parameters being passed to construct construct CommentDataPackage. Think about  creating a custom type that could potentially encapsulate some of the parameters here. Feel free to ping me if you want to discuss this offline. Also DEFAULT_MAX_COMMENTS is a constant. This doesn't need to be a class member. You can make this public and access it from CommentDataPackage.

**Review by suparnabarua — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 7 above)
Comment (KathyFeiyang) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Got it! I'll switch to isEmpty(). It can only be a empty string.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 8 above)
Comment (KathyFeiyang) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Got it! This was a typo. I'm sorry!

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 9 above)
Comment (KathyFeiyang) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Gotcha! I'll redesign this. Thanks!

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 7 above)
Comment (KathyFeiyang) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: https://github.com/KathyFeiyang/step/pull/8#discussion_r435939281: I couldn't directly respond to that question so I'll post it here:

The overriding is intentional, because there might be situations where maxCommentsToDisplay=0 but the page ID user input is invalid. The code outside of the maxCommentsToDisplay=0 case checks for invalid page ID. But I agree that the logic is hard to read. It was a bit hard to write as well. I will recheck the code to fix any issues and also clean it up.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 10]
```diff
@@ -82,36 +84,65 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
     int totalComments = commentHistoryList.size();
-    int totalPages;
 
     // Obtain the maximum number of comments to display. If the number is invalid, take the default value.
+    String maxCommentsToDisplayStr = request.getParameter("maxCommentsToDisplay");
     int maxCommentsToDisplay;
     try {
-      maxCommentsToDisplay = Integer.parseInt(request.getParameter("maxCommentsToDisplay"));
+      maxCommentsToDisplay = Integer.parseInt(maxCommentsToDisplayStr);
       if (maxCommentsToDisplay < 0) {
```
Comment (kellyktan) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Should this be < 1?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 11]
```diff
@@ -82,36 +84,65 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
     PreparedQuery commentHistory = datastore.prepare(commentHistoryQuery);
     List<Entity> commentHistoryList = commentHistory.asList(FetchOptions.Builder.withDefaults());
     int totalComments = commentHistoryList.size();
-    int totalPages;
 
     // Obtain the maximum number of comments to display. If the number is invalid, take the default value.
```
Comment (kellyktan) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: This function is getting a bit long. Would you be able to break each of these logical sections up into helper functions (with relevant names)?

File: portfolio/src/main/webapp/style.css [hunk 12]
```diff
@@ -87,8 +87,11 @@ nav,
 }
 
 /* Warning */
+#invalid-max-comments,
```
Comment (kellyktan) [inline portfolio/src/main/webapp/style.css]: Maybe instead of having separate ids for each, you could just have an "error" style class that you use for all of them? (or are there other differences between these that I'm not seeing?)

File: portfolio/src/main/webapp/index.html (hunk 2 above)
Comment (kellyktan) [reply portfolio/src/main/webapp/index.html]: Ooo great! I wonder if it would make sense to show the max number of pages (i.e. similar to "currently on page 1" you could do something like "page 1 of 10" or whatever)

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 10 above)
Comment (KathyFeiyang) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: The original behavior is that users can choose to not display any comments. Do you think I should remove the permission for maxCommentsToDisplay=0?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 11 above)
Comment (KathyFeiyang) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Yes, I'm working on it!

File: portfolio/src/main/webapp/style.css (hunk 12 above)
Comment (KathyFeiyang) [reply portfolio/src/main/webapp/style.css]: Got it! They look the same. I'll use one rule to add style to all of them

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 10 above)
Comment (kellyktan) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Ah gotcha.  I wonder if it'd be better to just include a way to toggle the whole comment section on/off instead?  (if that's too hard, feel free to leave as is)

Comment (KathyFeiyang) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Got it! I think that's doable. I'll try it. Then I'll remove the permission for 0 comments and later add a on/off checkbox or something

**Review by kellyktan — APPROVED** [review]
Looks great, thanks for making all of these changes!

**Comment by KathyFeiyang** [comment]
> Can you declare Usercomment right before invalidInputFlags? I think it's better to group all the primitive type declarations, followed by the other classes.

Of course! Just to double-check with you, my original code doesn't declare UserComment as an inner class because I thought UserComment might be used in other contexts. Should I make UserComment an inner class like the rest of the helper classes?

**Review by suparnabarua — APPROVED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 10 above)
Comment (KathyFeiyang) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: I added the on/off checkbox feature. It behaves as the following:
- When users check the checkbox to disable the comment history section, the entire comment history section disappears while everything that was displayed gets backed up (eg. the 2nd page, with 3 comments).
- When users uncheck the checkbox, the comment history section is restored to what it looked like before. Users can adjust the number of comments to show per page and the page ID, in the same way as before.

File: portfolio/src/main/webapp/script.js [hunk 13]
```diff
@@ -34,20 +38,62 @@ function addCyclicGreeting() {
...
 
 /**
- * Fetches and adds a history of comments, and the theoretical maximum and default
- * number of comments, to the page.
+ * Turns on (off) the comment history section if the current setting is
+ * off (on), and reload the comments (either removing or restoring the
+ * comments).
  */
-async function addComments() {
+async function turnOnOffCommentHistorySection() {
```
Comment (kellyktan) [inline portfolio/src/main/webapp/script.js]: I'd maybe call this "toggleCommentHistorySection()"

**Review by kellyktan — APPROVED** [review]
Looks great!

File: portfolio/src/main/webapp/script.js (hunk 13 above)
Comment (KathyFeiyang) [reply portfolio/src/main/webapp/script.js]: Gotcha! I'll change it. Thank you
